In [1]:
%matplotlib inline

### Importing Numpy, Keras and its optimizers

In [3]:
from __future__ import division,print_function
import numpy as np
from numpy.random import random

import keras
from keras.models import Sequential
from keras.layers.core import Flatten, Dense, Dropout, Lambda
from keras.optimizers import SGD

Using Theano backend.


In [37]:
x = random((30,1))

 ## y = 2 * x + 2


In [38]:
y = np.dot(x, 2.) + 2.

#### Showing Values of X

In [39]:
x

array([[ 0.1473186 ],
       [ 0.9052007 ],
       [ 0.44666439],
       [ 0.15512029],
       [ 0.49625389],
       [ 0.88094727],
       [ 0.95350086],
       [ 0.97620024],
       [ 0.07767123],
       [ 0.20444086],
       [ 0.4311851 ],
       [ 0.36499696],
       [ 0.50469161],
       [ 0.39442252],
       [ 0.68550133],
       [ 0.29571877],
       [ 0.20622951],
       [ 0.01471398],
       [ 0.32523735],
       [ 0.85129617],
       [ 0.33465863],
       [ 0.9128122 ],
       [ 0.9291631 ],
       [ 0.94645524],
       [ 0.31623003],
       [ 0.78733196],
       [ 0.37291789],
       [ 0.46133165],
       [ 0.94367175],
       [ 0.34948461]])

#### Showing Values of Y

In [40]:
y

array([[ 2.2946372 ],
       [ 3.81040141],
       [ 2.89332878],
       [ 2.31024058],
       [ 2.99250777],
       [ 3.76189455],
       [ 3.90700173],
       [ 3.95240048],
       [ 2.15534246],
       [ 2.40888171],
       [ 2.86237019],
       [ 2.72999391],
       [ 3.00938323],
       [ 2.78884503],
       [ 3.37100265],
       [ 2.59143755],
       [ 2.41245903],
       [ 2.02942796],
       [ 2.65047469],
       [ 3.70259234],
       [ 2.66931727],
       [ 3.8256244 ],
       [ 3.85832621],
       [ 3.89291048],
       [ 2.63246007],
       [ 3.57466393],
       [ 2.74583579],
       [ 2.92266329],
       [ 3.88734349],
       [ 2.69896921]])

#### In Keras Sequential Means a Model which is like seqeuntial stack of layers. Its just something Keras uses to create models.

#### In addition to this Keras, uses Dense method to create a linear model, which is what we are going to use in our linear problem

In [41]:
linear_model = Sequential([ Dense(1, input_shape=(1,)) ])

Here the first argument of Dense is the required  output dimension we required, here as the output is 1 dimensional array, we use 1. And as the input is also 1 dimensional we use 1

Then we call Compile method to the linear model we created with required Gradient descent Algorithm, Learning Rate and Loss function

Here, SGD is used for optimizer with learning rate of 0.1 and MSE or Mean Squared Loss as Loss function

In [42]:
linear_model.compile(optimizer=SGD(lr=0.1), loss='mse')

This compile method first initializes weights with random values which is far from correct values

If we evaluate or check the Y predictions now with the random values, we will get result of Loss Function very large

In [43]:
linear_model.evaluate(x, y)

30/30 [==============================] - 0s


7.3708395957946777

To calculate weights with given input/output pair, we have to call fit() method on our model

In [44]:
linear_model.fit(x,y, nb_epoch=5, batch_size=1)

Epoch 1/5
30/30 [==============================] - 0s - loss: 0.5235     
Epoch 2/5
30/30 [==============================] - 0s - loss: 0.0100     
Epoch 3/5
30/30 [==============================] - 0s - loss: 0.0035     
Epoch 4/5
30/30 [==============================] - 0s - loss: 0.0017         
Epoch 5/5
30/30 [==============================] - 0s - loss: 6.4502e-04 


Here nb_epoch is the number of epochs the calculation/learning should be run. 1 Epoch means going through each input ones... so 5 Epoch means going through all the inputs 5 times. 

Generally input data is fed in batches/collection, which are like 10-100 input data at onces, not all of them at once

Now if we check the loss value, it will be lot less than before

In [45]:
linear_model.evaluate(x, y, verbose=0)

0.0003556846349965781

We can only check the weights, the model has calculated

In [46]:
linear_model.get_weights()

[array([[ 1.94116473]], dtype=float32), array([ 2.03684092], dtype=float32)]

here 1.9411 is nearly equal to correct value which is 2, and the bias 2.0368 which is close to 2.0